src/main/java/picard/sam/SAMSplitter.java

```diff
@@ -0,0 +1,183 @@
+ * multiple BAM files of approximately equal size. This will retain the sort order
+ * within each output BAM and if the BAMs are concatenated in order (output files are named
+ * numerically) the order of the reads will match the original BAM. As this tool is
+ * intended for unmapped input files, all secondary and supplementary reads will be removed.
+ */
+@CommandLineProgramProperties(
+        summary = SAMSplitter.USAGE_SUMMARY + SAMSplitter.USAGE_DETAILS,
+        oneLineSummary = SAMSplitter.USAGE_SUMMARY,
+        programGroup = SamOrBam.class)
+@DocumentedFeature
+public class SAMSplitter extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Splits a SAM or BAM file to multiple BAMs.";
+    static final String USAGE_DETAILS = "This tool splits the input SAM/BAM file into multiple BAM files " +
+            "This can be used to split a large unmapped BAM in order to parallelize alignment."+
+            "<br />" +
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar SAMSplitter \\<br />" +
+            "     I=paired_unmapped_input.bam \\<br />" +
+            "     OUTPUT_DIR=out_dir \\ <br />" +
+            "     TOTAL_READS_IN_INPUT=800000000" +
+            "     SPLIT_TO_N_TEMPLATES=24000000" +
+            "</pre>" +
+            "<hr />";
+    @Argument(doc = "Input SAM/BAM file to split", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Argument(shortName = "N", doc = "Split to have approximately N templates per output file.")
+    public int SPLIT_TO_N_TEMPLATES;
+
+    @Argument(shortName = "TOTAL_READS", doc = "Total number of reads in the input file.")
+    public int TOTAL_READS_IN_INPUT;
+
+    @Argument(shortName = "ODIR", doc = "Directory in which to output the split BAM files.")
+    public File OUTPUT_DIR;
+
+    @Argument(shortName = "PAIR", doc = "If true the input bam contains paired reads and " +
+            "each template is assumed to contain two reads (rather than one for unpaired inputs).")
+    public boolean PAIRED_INPUT = true;
+
+    private final Log log = Log.getInstance(SAMSplitter.class);
+
+    public static void main(final String[] argv) {
+        System.exit(new SAMSplitter().instanceMain(argv));
+    }
+
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+        final SAMFileHeader header = reader.getFileHeader();
+        if (header.getSortOrder() == SAMFileHeader.SortOrder.coordinate) {
+            throw new PicardException("Cannot split coordinate sorted bam. Only queryname sorted (or querygrouped) " +
+                    "input will ensure that templates are not split into multiple files.");
+        }
+
+        final Map<String, SAMRecord> firstSeenMates = new HashMap<>();
```

Comment:
`firstSeenMates` is not an intuitive variable name.  Perhaps `readPairCache`? 

Also, I don't see why you need this if you assume the input is queryname or query-grouped?